rainier-core/src/main/scala/com/stripe/rainier/sampler/LeapFrog.scala

```diff
@@ -1,23 +1,77 @@
 package com.stripe.rainier.sampler
 
-private[sampler] case class LeapFrog(
-    nVars: Int,
-    initialHalfThenFullStep: Array[Double] => Unit,
-    twoFullSteps: Array[Double] => Unit,
-    finalHalfStep: Array[Double] => Unit
-) {
+private[sampler] case class LeapFrog(density: DensityFunction) {
   /*
   Params layout:
   array(0..(n-1)) == ps
   array(n..(n*2-1)) == qs
   array(n*2) == potential
   array(n*2+1) == stepSize
    */
+  private val nVars = density.nVars
   private val potentialIndex = nVars * 2
   private val stepSizeIndex = potentialIndex + 1
   private val inputOutputSize = stepSizeIndex + 1
   private val buf1 = new Array[Double](inputOutputSize)
 
+  def halfPs(inputs: Array[Double]): Unit = {
+    density.update(inputs)
```

Comment:
this won't work right because the ps are at the start of the array, right?

Reply:
Ah, right, I was imaging the DensityFunction could have the whole big array, but it should definitely only know about the qs.

Reply:
I think this is good now.